bugbear.py

```diff
@@ -778,6 +740,79 @@ def visit(self, node):
         return node
 
 
+class FuntionDefDefaultsVisitor(ast.NodeVisitor):
+    def __init__(self, b008_extend_immutable_calls=None):
+        self.b008_extend_immutable_calls = b008_extend_immutable_calls or set()
+        for node in B006.mutable_literals + B006.mutable_comprehensions:
+            setattr(self, f"visit_{node}", self.visit_mutable_literal_or_comprehension)
+        self.errors = []
+        self.arg_depth = 0
+        super().__init__()
+
+    def visit_mutable_literal_or_comprehension(self, node):
+        # Flag B006 iff mutable literal/comprehension is not nested.
+        # We only flag these at the top level of the expression as we
+        # cannot easily guarantee that nested mutable structures are not
+        # made immutable by outer operations, so we prefer no false positives.
+        # e.g.
+        # >>> def this_is_fine(a=frozenset({"a", "b", "c"})): ...
+        #
+        # >>> def this_is_not_fine_but_hard_to_detect(a=(lambda x: x)([1, 2, 3]))
+        #
+        # We do still search for cases of B008 within mutable structures though.
+        if self.arg_depth == 1:
+            self.errors.append(B006(node.lineno, node.col_offset))
+        # Check for nested functions.
+        self.generic_visit(node)
+
+    def visit_Call(self, node):
+        call_path = ".".join(compose_call_path(node.func))
+        if call_path in B006.mutable_calls:
+            self.errors.append(B006(node.lineno, node.col_offset))
+        elif call_path not in B008.immutable_calls | self.b008_extend_immutable_calls:
+            # Check if function call is actually a float infinity/NaN literal
+            if call_path == "float" and len(node.args) == 1:
+                float_arg = node.args[0]
+                if sys.version_info < (3, 8, 0):
+                    # NOTE: pre-3.8, string literals are represented with ast.Str
+                    if isinstance(float_arg, ast.Str):
+                        str_val = float_arg.s
+                    else:
+                        str_val = ""
+                else:
+                    # NOTE: post-3.8, string literals are represented with ast.Constant
+                    if isinstance(float_arg, ast.Constant):
+                        str_val = float_arg.value
+                        if not isinstance(str_val, str):
+                            str_val = ""
+                    else:
+                        str_val = ""
+
+                # NOTE: regex derived from documentation at:
+                # https://docs.python.org/3/library/functions.html#float
+                inf_nan_regex = r"^[+-]?(inf|infinity|nan)$"
+                re_result = re.search(inf_nan_regex, str_val.lower())
+                is_float_literal = re_result is not None
+            else:
+                is_float_literal = False
+
+            if not is_float_literal:
+                self.errors.append(B008(node.lineno, node.col_offset))
```

Comment:
I know this was previously existing code, but we can shorten it *considerably* by using `ast.literal_eval()`:

```suggestion
                try:
                    value = float(ast.literal_eval(node.args[0]))
                except Exception:
                    pass
                else:
                    if math.isfinite(value):
                        self.errors.append(B008(node.lineno, node.col_offset))
```

(OK, I admit that this no longer warns on a redundant `float()` call wrapped around an "infinity literal" like `1e999`, but IMO that's fine.)

Reply:
Thanks! I've updated the code accordingly and it looks a lot nicer now 😄